### ISSUE TITLE
bug: Argo Workflow Status is overwritten with an empty status in the WEC

### ISSUE THREAD

**Issue by pdettori** [kubestellar/kubestellar#2077]
### Describe the bug

This bug happens when using KubeFlow Pipelines with KubeStellar.  A KubeFlow Pipeline run generates an Argo Workflow with immutable spec. Only status gets updated by the argo workflow controller during workflow execution. The Argo Workflow API resource does not use a sub-resource for status. KubeStellar currently cleans the status for all objects. As a result, the local status in the WEC gets overwritten with an empty status.

In the current tests, the status add-on up-syncs the non-empty status to the ITS (from which it is synced to the WDS source object). Status add-on does not sync an empty status, therefore the status on the WDS object is the status on the WEC before it gets overwritten to empty by the work-agent sync.

Here is the result:

As soon as the workflow is applied on the WDS, both the object on the WDS and the WEC have non-empty status

```shell
$ kubectl get workflow -n kubeflow
NAME                          STATUS    AGE
tutorial-data-passing-75zsd   Running   30s
$ kubectl get workflow -n kubeflow --context cluster2
NAME                          STATUS    AGE
tutorial-data-passing-75zsd   Running   40s
```

after workflow completes, the workflow on the WDS has a completed status and the one on the WEC has an empty status:

```shell
$ kubectl get workflow -n kubeflow
NAME                          STATUS      AGE
tutorial-data-passing-75zsd   Succeeded   9m15s
$ kubectl get workflow -n kubeflow --context cluster2
NAME                          STATUS   AGE
tutorial-data-passing-75zsd            
```

Status is completely empty on the WEC:

```shell
$ kubectl get workflow -n kubeflow --context cluster2 tutorial-data-passing-75zsd -o=jsonpath='{.status}'
<NOTHING SHOWN>
```

While on WDS:

```shell
$ kubectl get workflow -n kubeflow  tutorial-data-passing-75zsd -o=jsonpath='{.status}'
{"artifactRepositoryRef":{"artifactRepository":{"archiveLogs":true,"s3":{"accessKeySecret":{"key":"accesskey","name":"mlpipeline-minio-artifact"},"bucket":"mlpipeline","
.....
```

### Steps To Reproduce

1. Install KubeStellar from main
2.  Create an ITS of type host (no -p)
3. Deploy  OCM directly in the hosting cluster
4. Create a WDS of type host with kubestellar and transport controllers
5. Install Argo Workflows on both the WDS and the WECs
6. Give RBAC permissions for argo workflows to the OCM klusterlet
7. Create a binding policy with singleton status set 
8.  Create a workflow on WDS with suspend set to true
9. Observe how initially status gets updated on WDS workload object, and at some point only the WDS object has a status while the WEC object has empty status.

### Expected Behavior

It should not overwrite workflow object on WEC after creation.

### Additional Context

Since the Workflow  spec created by KFP is assumed to be immutable, a possible solution would be to re-introduce support for making auto-downsync optional as we had in KS 0.14 See https://github.com/kubestellar/kubestellar/pull/1099. OCM ManifestWork has a `updateStrategy` option that allows to specify `CreateOnly`.

**Comment by pdettori** [comment]
One key difference from the approach taken ij #1099 is that since a ManifestWork may wrap multiple objects, simply labeling each object to mark for `CreateOnly` may lead to conflict (what if some objects do not need the `CreateOnly` and some do ?). Therefore, since currently BindingPolicy maps to a ManifestWork, a new flag should be introduced on `BindingPolicy`. 

cc: 
@MikeSpreitzer
@ezrasilvera 
@vMaroon 
@nirrozenbaum

**Comment by MikeSpreitzer** [comment]
I am having trouble understanding the words describing what is happening.

> Status add-on does not sync an empty status

Does that mean (a) when the status in the WEC is empty, the status add-on does nothing or (b) the status that the status add-on writes into the ITS is not empty?

**Comment by MikeSpreitzer** [comment]
> As soon as the workflow is applied on the WDS, both the object on the WDS and the WEC have non-empty status

"As soon as the workflow is applied on the WDS", it exists nowhere else. It is only after some time has passed and additional steps have been takend that the workflow exists elsewhere. I am struggling to understand what you are saying about that intermediate stretch of time.

**Comment by MikeSpreitzer** [comment]
Can ManifestWork do what we need here? I have always expected that we will eventually have to shift to using the policy arm of OCM rather than the application arm, in order to get all the functionality that we want.

**Comment by MikeSpreitzer** [comment]
What we need in this situation is not the "create only" flag of ManifestWork; that flag prevents propagation of all changes from WDS to WEC. What we need is a way to say that the status should not propagate downward but the other content _should_.

**Comment by MikeSpreitzer** [comment]
> One key difference from the approach taken ij https://github.com/kubestellar/kubestellar/pull/1099 is that since a ManifestWork may wrap multiple objects, simply labeling each object to mark for CreateOnly may lead to conflict (what if some objects do not need the CreateOnly and some do ?).

I do not follow. In ManifestWork, each of the wrapped objects individually has its own ManifestConfigOption, which is where the UpdateStrategy (which may be "Update", "CreateOnly", or "ServerSideApply") appears.

**Comment by MikeSpreitzer** [comment]
Using the UpdateStrategy "ServerSideApply" rather than "CreateOnly" would mean that adds and changes propagate downward --- but content removals do not.

**Comment by pdettori** [comment]
@MikeSpreitzer: some answers here:

> Does that mean (a) when the status in the WEC is empty, the status add-on does nothing or (b) the status that the status add-on writes into the ITS is not empty?

It is (a). The status-add on makes only updates for a non-empty status. The workstatus for the workflow in my scenario shows the last status before the workflow status n the WEC was overwritten by an empty status.

> It is only after some time has passed and additional steps have been takend that the workflow exists elsewhere. I am struggling to understand what you are saying about that intermediate stretch of time.

I did not elaborate all the steps. At high level, the workflow is applied on the WDS with `.spec.suspend=true` and a label that match a binding policy with a target WEC `cluster1`.  KS (Binding controller + OTS + OCM) sync the workflow  to the WEC. During this process KS (with my custom code for now, with custom transform later) filters out the `suspend` flag. The Argo Workflow Controller runs the Workflow on the WEC, and the status add-on syncs back the status to the ITS. The Status controller syncs the status from the ITS to the WDS.  Sometimes during this process, the OCM Work Agent syncs the Workflow (with an empty status) from the manifestwork and the status of the workflow in the wec becomes empty.

**Comment by pdettori** [comment]
> I do not follow. In ManifestWork, each of the wrapped objects individually has its own ManifestConfigOption, which is where the UpdateStrategy (which may be "Update", "CreateOnly", or "ServerSideApply") appears.

Yes, you are correct, that should work with the current design. Will check if "ServerSideApply" works for this use case.

**Comment by pdettori** [comment]
Did a quick test stopping the transport controller and manually editing the ManifestWork to add the configuration for `ServerSideApply` - it seems to work. As soon I restarted the transport controller it set the `manifestwork` back to default and the status disappeared from the Workload object in the WEC.

**Comment by ezrasilvera** [comment]
@pdettori  (cc: @MikeSpreitzer ) I'm not sure that ServerSideApply  solves the issue of labeling a strategy for ManifestWork while individual objects within the ManifestWork  may require different strategy. The specific KubeFlow use case may not require such granularity, however in general cases shouldn't we expect objects that do want removal updates for some objects while avoid this for other objects ?

**Comment by pdettori** [comment]
@ezrasilvera the Manifest Work update strategy is per object, not per manifest. So you can have different objects with different update strategies in the same manifest.

**Comment by ezrasilvera** [comment]
> @ezrasilvera the Manifest Work update strategy is per object, not per manifest. So you can have different objects with different update strategies in the same manifest.

Thanks for the clarification

**Comment by clubanderson** [comment]
Issues go stale after 90d of inactivity.
After a furter 30 days, they will turn rotten.
Mark the issue as fresh with `/remove-lifecycle stale`.

If this issue is safe to close now please do so with `/close`.

/lifecycle stale

**Comment by clubanderson** [comment]
Stale issues rot after 30d of inactivity.
Mark the issue as fresh with `/remove-lifecycle rotten`.
Rotten issues close after an additional 30d of inactivity.

If this issue is safe to close now please do so with `/close`.

/lifecycle rotten

**Comment by clubanderson** [comment]
Rotten issues close after 30d of inactivity.
Reopen the issue with `/reopen`.
Mark the issue as fresh with `/remove-lifecycle rotten`.

/close
